### PR TITLE
Move build system over to custom runner

### PR DESCRIPTION
**Pull request by DRMacIver**
This is a start of trying to unify our build system a bit more so that it's less spread over Python/make/shell/etc. There's still a bit of shell (both for bootstrapping into Python and some of our test scripts remain).

The main goal in doing this is that it will make it easier to refactor our decisions about what should be running. Right now this is mostly feature parity with a few things that got fixed while I was in the area (e.g. we now check a few more RST files).

I'm not super happy with this code, but it's a decent starting point.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
+# This script is here to bootstrap the Hypothesis build process into a working
+# version of Python, then hand over to the actual Hypothesi build runner (which
```
Comment (alexwlchan) [inline build.sh]: nit: Hypothesi

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# is written in Python instead of bash).
+
+set -e -u -x
```
Comment (alexwlchan) [inline build.sh]: minor: I prefer the verbose forms (e.g. `set -o errexit`) for readability.

**Comment by DRMacIver** [comment]
It looks like the remaining problems are about how tox looks up interpreters. I'm not going to fix them tonight.

I'd been eyeing up killing tox as part of this and I think that may have tipped me over towards doing that.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# You don't need to use this Makefile and should use build.sh instead. This is
+# just here so that us poor souls who remember the Make based system and keep
+# typing "make target" can ease our transition to the new sytsem.
```
Comment (Zac-HD) [inline Makefile]: sytsem -> system

(you poor soul :wink:)

File: hypothesis-python/scripts/basic-test.sh [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -e -o xtrace
 
-cd "$(dirname "$0")"/../hypothesis-python
+cd "$(dirname "$0")"/..
```
Comment (alexwlchan) [inline hypothesis-python/scripts/basic-test.sh]: Is this meant to be cd'ing to the directory this script lives in?

If so, would https://stackoverflow.com/a/246128/1558022 potentially be more robust? (e.g. if this script is called from something else, is `$0` set correctly?)

File: hypothesis-python/tox.ini [hunk 5]
```diff
@@ -146,7 +146,7 @@ commands =
     python -m coverage debug sys
     python -m coverage run --rcfile=.coveragerc -m pytest -n0 --strict tests/cover tests/datetime tests/py3 tests/numpy tests/pandas --maxfail=1 --ff {posargs}
     python -m coverage report -m --fail-under=100 --show-missing
-    python ../scripts/validate_branch_check.py
+    python ./scripts/validate_branch_check.py
```
Comment (alexwlchan) [inline hypothesis-python/tox.ini]: nit: why the leading `./` here and not on line 60?

File: tooling/scripts/ensure-python.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+set -o nounset
+set -x
+
+# shellcheck source=tooling/scripts/common.sh
+source $(dirname "$0")/common.sh
+
+# This is to guard against multiple builds in parallel. The various installers will tend
+# to stomp all over each other if you do this and they haven't previously successfully
+# succeeded. We use a lock file to block progress so only one install runs at a time.
+# This script should be pretty fast once files are cached, so the lost of concurrency
```
Comment (alexwlchan) [inline tooling/scripts/ensure-python.sh]: nit: "lost" ~> "loss

File: tooling/scripts/ensure-python.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+# is not a major problem.
+# This should be using the lockfile command, but that's not available on the
+# containerized travis and we can't install it without sudo.
+# Is is unclear if this is actually useful. I was seeing behaviour that suggested
```
Comment (alexwlchan) [inline tooling/scripts/ensure-python.sh]: nit: "It is unclear"

File: tooling/scripts/ensure-python.sh [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+# concurrent runs of the installer, but I can't seem to find any evidence of this lock
+# ever not being acquired.
```
Comment (alexwlchan) [inline tooling/scripts/ensure-python.sh]: Surely a single Travis run only installs a single version of Python, in which case parallelism is never a problem. Is this only an issue for local builds?

File: tooling/src/hypothesistooling/__init__.py [hunk 9]
```diff
@@ -368,9 +375,9 @@ def changed_files_from_master():
     return files
 
 
-def should_run_ci_task(task, is_pull_request):
+def should_run_ci_task(task):
     """Given a task name, should we run this task?"""
-    if not is_pull_request:
+    if not IS_PULL_REQUEST:
```
Comment (alexwlchan) [inline tooling/src/hypothesistooling/__init__.py]: 👍 

much neater

**Comment by DRMacIver** [comment]
BTW the current status of this is that I'm struggling to reproduce the problem that is occurring reliably on CI where the install phase isn't actually working, and it's a rather slow loop to reproduce so I keep getting distracted by other work. Still ongoing work in progress though!

**Comment by DRMacIver** [comment]
> BTW the current status of this is that I'm struggling to reproduce the problem that is occurring reliably on CI 

`(╯°□°）╯︵ ┻━┻`

Eventually I figured out by staring at CI logs for long enough.

The problem is that when running a script that exists solely to read stuff out of the environment, *maybe I shouldn't be carefully starting it in a blank environment*. Ugh.

File: tooling/src/hypothesistooling/__main__.py [hunk 10]
```diff
@@ -45,10 +45,11 @@ def task(fn):
 
 @task
 def lint():
-    os.chdir(tools.HYPOTHESIS_PYTHON)
-    pip_tool('flake8', 'src', 'tests', '--config', os.path.join(
-        tools.ROOT, '.flake8'
-    ))
+    pip_tool(
+        'flake8',
```
Comment (Zac-HD) [inline tooling/src/hypothesistooling/__main__.py]: Note: our current run requires `flake8-docstrings` to be installed too.

Comment (DRMacIver) [reply tooling/src/hypothesistooling/__main__.py]: That's fine. This runs in a virtualenv that installs everything from `tools.txt`

**Comment by Zac-HD** [comment]
Looks like all you need to do is drop the old-python and old-pip tasks from Travis :smile:

**Comment by DRMacIver** [comment]
Have fixed all of the small nits, but I think this is all good for review now.

**Comment by alexwlchan** [comment]
I have seen this, but I don’t have a lot of free time for OSS work right now, and I’m on holiday for a week from Sunday. I’d be surprised if I manage to review this properly within the next fortnight.

I won’t be upset if you want to merge this before I’ve reviewed it, and we patch it up after the fact.

File: hypothesis-python/scripts/basic-test.sh (hunk 4 above)
Comment (dchudz) [reply hypothesis-python/scripts/basic-test.sh]: @DRMacIver the 👍 makes me think you wanted to take this suggestion, but it looks like you didn't?

(I don't have any opinion and definitely haven't read that SO answer. Just checking in to see if there's something you wanted to do and didn't here.)

Comment (DRMacIver) [reply hypothesis-python/scripts/basic-test.sh]: Whoops! I did it in a bunch of places but not the one that @alexwlchan actually commented on. Thanks.

**Comment by DRMacIver** [comment]
> I won’t be upset if you want to merge this before I’ve reviewed it, and we patch it up after the fact.

No worries. I'll definitely want to merge this in the next fortnight, but this is very much a stage one so it will probably see a lot of evolution from here, so you'll have plenty of later opportunity to comment. :-)

@froztbyte any chance you could take a look at this? It's rather in your area of expertise.

**Comment by DRMacIver** [comment]
> just don't have the depth of experience with bash to do a useful review.

FWIW very little of the bash is actually new. It's all just extracted from existing bash scripts that have been more or less working for about four years now. That doesn't mean it's *correct*, but it's probably not broken in new ways, and it doesn't need a super extensive review.

File: build.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+set -o errexit
+set -o nounset
+
+ROOT="$(git -C "$(dirname "$0")" rev-parse --show-toplevel)"
+
+export HYPOTHESIS_ROOT="$ROOT"
+
+SCRIPTS="$ROOT/tooling/scripts"
+
+# shellcheck source=tooling/scripts/common.sh
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply build.sh]: It's not a command, it's a shellcheck pragma to tell it where to look for the dynamic source.

File: tooling/scripts/common.sh [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+# scripts so that they can share a common set of paths conveniently.
+
+set -o errexit
+set -o nounset
+
+HERE="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+ROOT="$(git -C "$HERE" rev-parse --show-toplevel)"
+
+export ROOT
+export BUILD_RUNTIMES=${BUILD_RUNTIMES-$HOME/.cache/hypothesis-build-runtimes}
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply tooling/scripts/common.sh]: Both are valid syntax! Though it turns out with very slightly different semantics - `:-` will use the new value if `VARIABLE` is set but null, while `-` will use null in that case. I'll switch it over.

[the remark this answers is not in the document]
Comment (alexwlchan) [reply tooling/scripts/common.sh]: Both “Why bash? 😭” and “Why? bash 😭” are valid responses I had! Though it turns out with very slightly different semantics.



😉

File: build.sh [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+TOOL_HASH=$("$PYTHON" "$SCRIPTS/tool-hash.py" < "$TOOL_REQUIREMENTS")
+
+TOOL_VIRTUALENV="$VIRTUALENVS/build-$TOOL_HASH"
+TOOL_PYTHON="$TOOL_VIRTUALENV/bin/python"
+
+if [ ! -e "$TOOL_PYTHON" ] ; then
+    rm -rf "$TOOL_VIRTUALENV"
+    "$PYTHON" -m pip install --upgrade virtualenv
+	"$PYTHON" -m virtualenv "$TOOL_VIRTUALENV"
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply build.sh]: Turns out to be mixed tabs and spaces from moving some stuff over from the makefile. Whoops!

**Comment by DRMacIver** [comment]
BTW I propose that at some point we just YOLO merge this rather than doing a detailed review.

Reasons:

* It's large and annoying. I know, sorry.
* A detailed review will conclude that it's not very good, but we know that already
* Most of the details will change in smaller chunks as we monorepo more stuff, and *that* can be subject to more detailed review
* It doesn't actually affect users because it doesn't touch any shippable code, so any problems with it will fall on us to fix later rather than inflicting them on other people.

**Comment by DRMacIver** [comment]
> I expect a follow-up issue to track the inevitable glitches, but let's get it in and iterate on a shared base.

Oh for sure. But I mean hey things are still broken from the *last* big change, so how much worse can it be? 😉

**Comment by Zac-HD** [comment]
Things *including certain build tasks* are still broken, so... :wink:

**Comment by DRMacIver** [comment]
> Things including certain build tasks are still broken, so... 😉

Specifically there are several task that are broken in master but work correctly in this branch and so are picking up failures that are also present in master. 😉

**Comment by Zac-HD** [comment]
Merge when you think it's ready!

**Comment by alexwlchan** [comment]
> BTW I propose that at some point we just YOLO merge this rather than doing a detailed review.

What Could Possibly Go Wrong.

**Comment by DRMacIver** [comment]
> What Could Possibly Go Wrong.

Literally Everything.

**Pull request merged**